rtmr/extend/extend.go

```diff
@@ -0,0 +1,199 @@
+		return fmt.Errorf("could not read  %s: %v", TsmPathTcgMap, err)
+	}
+	var rtmrPcrMaps = map[int]string{
+		0: "1,7\n",
+		1: "2-6\n",
+		2: "8-15\n",
+		3: "\n",
+	}
+
+	inputMap := convertCstring(data)
+	if inputMap != rtmrPcrMaps[r.RtmrIndex] {
+		return fmt.Errorf("tcg map error. expect:%s get:%s", rtmrPcrMaps[r.RtmrIndex], string(data))
+	}
+	return nil
+}
+
+// SearchRtmrInterface searches for an rtmr entry in the configfs.
+func SearchRtmrInterface(client configfsi.Client, index int) (*RtmrExtend, error) {
+	root := TsmRtmrPrefix
+	out := RtmrExtend{}
+	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			p, err := configfsi.ParseTsmPath(path)
+			if err != nil {
+				return err
+			}
+			r := RtmrExtend{
+				RtmrIndex: index,
+				entry:     &configfsi.TsmPath{Subsystem: RtmrSubsystem, Entry: p.Entry},
+				client:    client,
+			}
+			if r.ValidateRtmrIndex() == nil {
+				out = r
+				return nil
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	if out.ValidateRtmrIndex() != nil {
+		return nil, fmt.Errorf("could not find rtmr entry in configfs: %v", err)
+	}
+
+	return &out, nil
+}
+
+// CreateRtmrInterface creates a new rtmr entry in the configfs.
+func CreateRtmrInterface(client configfsi.Client, index int) (r *RtmrExtend, err error) {
+	entryPath, err := client.MkdirTemp(TsmRtmrPrefix, fmt.Sprintf("rtmr%d-", index))
+	if err != nil {
+		return nil, fmt.Errorf("could not create rtmr entry in configfs: %v", err)
+	}
+	p, _ := configfsi.ParseTsmPath(entryPath)
+
+	r = &RtmrExtend{
+		RtmrIndex: index,
+		entry:     &configfsi.TsmPath{Subsystem: RtmrSubsystem, Entry: p.Entry},
+		client:    client,
+	}
+
+	if err := r.SetRtmrIndex(); err != nil {
+		return nil, fmt.Errorf("could not set rtmr index %d: %v", index, err)
+	}
+	return r, nil
+}
+
+// ExtendtoRtmrClient extends the measurement to the rtmr with the given client.
+func ExtendtoRtmrClient(client configfsi.Client, rtmr int, hashAlgo crypto.Hash, content []byte) error {
+	if hashAlgo != crypto.SHA384 {
+		return fmt.Errorf("unsupported hash algorithm %v", hashAlgo)
+	}
+	if len(content) == 0 {
+		return fmt.Errorf("input event log is empty")
+	}
+	if rtmr < 0 || rtmr > 3 {
+		return fmt.Errorf("invalid rtmr index %d. Index can only be 0-3", rtmr)
+	}
+
+	r, err := SearchRtmrInterface(client, rtmr)
+	if err != nil {
+		logger.V(2).Infof("Could not find rtmr entry in configfs. error: %v", err)
+		logger.V(1).Info("Creating new rtmr entry in configfs.")
+		r, err = CreateRtmrInterface(client, rtmr)
+		if err != nil {
+			return err
+		}
+	} else {
+		logger.V(1).Info("Found existing rtmr entry in configfs.")
+	}
+
+	hash := sha512.Sum384(content)
+	err = r.ExtendDigest(hash[:])
+	return err
+}
+
+// ExtendtoRtmr extends the measurement to the rtmr with the given hash algorithm and content.
+func ExtendtoRtmr(rtmr int, hashAlgo crypto.Hash, content []byte) error {
```

Comment:
ExtendtoRtmr is an odd name. ExtendRtmr is more natural. Do we think that requiring pre-hashed content is the right interface? I would have thought it should be a thin layer to the tsm subsystem, and you can add the extras in a usage library.

Reply:
Done. I feel that requiring pre-hashed content is the right interface. Here is the interface for tpm: https://github.com/google/go-tpm-tools/blob/264c9b4c927abb90293328c4e16f1e4251bc92dc/cel/canonical_eventlog.go#L138

Anyway I added a new thin function